From 1b165bd4c164803e35dac3e2d902d24a3e66a8d1 Mon Sep 17 00:00:00 2001
From: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Date: Tue, 7 Jul 2020 15:17:01 -0700
Subject: [PATCH 586/767] driver: serdes_debugfs: Allow user to clear prbs
 errors.

commit 5a00ca7e44e36351d2d399a53b8b3892cf89abc2 from
git@git.assembla.com:cavium/WindRiver.linux.git

Added new command to clear PRBS errors

Change-Id: I1ac5897ca0ef3b03ed819a7a74c1f53629d5dc41
Signed-off-by: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/31575
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../marvell/octeontx2-serdes/serdes_debugfs.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c b/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
index 81827966c09a..f5930a9c6bf1 100644
--- a/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
+++ b/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
@@ -54,7 +54,8 @@ struct dentry *pserdes_root;
 enum cgx_prbs_cmd {
 	CGX_PRBS_START_CMD = 1,
 	CGX_PRBS_STOP_CMD,
-	CGX_PRBS_GET_DATA_CMD
+	CGX_PRBS_GET_DATA_CMD,
+	CGX_PRBS_CLEAR_CMD
 };
 
 struct cgx_prbs_errors {
@@ -460,6 +461,8 @@ static int serdes_dbg_prbs_lane_parse(const char __user *buffer,
 					-EINVAL;
 		} else if (!strcmp(subtoken, "stop")) {
 			*cmd = CGX_PRBS_STOP_CMD;
+		} else if (!strcmp(subtoken, "clear")) {
+			*cmd = CGX_PRBS_CLEAR_CMD;
 		} else {
 			ec = -EINVAL;
 		}
@@ -557,6 +560,20 @@ static ssize_t serdes_dbg_prbs_write_op(struct file *filp,
 			pr_info("GSER PRBS stop on QLM %d on Lane %d.\n", qlm, qlm_lane);
 		break;
 
+	case CGX_PRBS_CLEAR_CMD:
+		arm_smccc_smc(OCTEONTX_SERDES_DBG_PRBS, cmd,
+			      qlm, 0, qlm_lane, 0, 0, 0, &res);
+		if (res.a0 != SMCCC_RET_SUCCESS) {
+			pr_info("GSER prbs clear command failed.\n");
+			return -EIO;
+		}
+		if (qlm_lane == -1)
+			pr_info("GSER PRBS errors cleared on QLM%d\n", qlm);
+		else
+			pr_info("GSER PRBS errors cleared on QLM%d Lane%d\n",
+					qlm, qlm_lane);
+		break;
+
 	default:
 		pr_info("GSER PRBS set QLM %d to read.\n", qlm);
 		break;
-- 
2.31.1

